presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -558,9 +558,9 @@ protected Type visitSearchedCaseExpression(SearchedCaseExpression node, Stackabl
         @Override
         protected Type visitSimpleCaseExpression(SimpleCaseExpression node, StackableAstVisitorContext<Context> context)
         {
-            for (WhenClause whenClause : node.getWhenClauses()) {
-                coerceToSingleType(context, whenClause, "CASE operand type does not match WHEN clause operand type: %s vs %s", node.getOperand(), whenClause.getOperand());
-            }
+            coerceToSingleType(context,
+                    "All CASE results must be the same type: %s",
```

Comment:
In this case, we want the error message to indicate *which* `when` operand cannot be compared to the `case` operand. So the fix is a little more involved. We need to attempt to find the common supertype between each `when` operand and the `case` operand and report an error if necessary. As we do this, keep track of the common supertype between all the types seen so far. Then, at the end, add a coercion for the case operand and all when operands to the common supertype. To illustrate:

```
for each when_operand in when_clauses
    t = supertype(when_operand, case_operand)
    if t not found => error
    common_type_so_far = supertype(t, common_type_so_far)   // or just t in the first loop

add_coercion(case_operand, common_type_so_far)
for each when_operand in when_clauses
    add_coercion(when_operand, common_type_so_far)
```

Reply:
it has been modified according to your suggestion